# P01 - Regresión

# Predicción y análisis de profit de ventas de bicicletas

* Andre Esteban
* Nicolas Martinez
* Gonzalo Cano

---

## Objetivos

**Objetivos generales:**

El objetivo general de este proyecto es hacer un análisis de los datos de las ventas de una empresa que se dedica a vender bicicletas mediante la aplicación de modelos de regresión y técnicas de regularización, con el fin de identificar qué factores son los más relevantes para las ventas de la empresa y proponer modelos que nos permitan optimizar la toma de decisiones comerciales para la empresa.

**Objetivos Específicos:**

Para poder cumplir el objetivo del proyecto lo que se tendrá que hacer es lo siguiente:

* Analizar el dataset de la empresa, identificando la estructura, que variables tenemos disponibles y la información relevante que aportan las muestras.

* Limpiar y transformar los datos, ajustando las variables necesarias para que puedan usarse adecuadamente en los modelos de regresión.

* Usar y comparar diferentes enfoques de regresión para evaluar el ajuste de los modelos a través de las métricas de desempeño.

* Utilizar métodos de penalización para evaluar el impacto en la predicción y en la significancia de las variables.

* Diseñar un Pipeline que incluya de forma estructurada todos los pasos que se realizaron para hacer el modelo.

* Evaluar los valores estadísticos de los modelos y compararlos para identificar aquel modelo que tenga un mejor balance.

* A partir de los resultados obtenidos concluir sobre la efectividad de los modelos y la relevancia de los factores en las ventas de la empresa.


---
## Marco Téorico

La regresión lineal es una herramienta estadística que predice o estima el valor de una variable, llamada dependiente, en función de una o más variables independientes. Según IBM, busca ajustar una línea recta, o plano, si hay múltiples variables, que minimice la diferencia entre los valores reales y los valores predichos, mediante un método de “mínimos cuadrados”. Con ello genera una fórmula matemática interpretable que explica la relación entre variables. Esta técnica es usada en muchos ámbitos (negocios, ciencias sociales, ambientales) porque permite transformar datos en pronósticos útiles y entender cómo cambian los resultados cuando varían las variables explicativas. (IBM, 2021)

$$ y = b_0 + b_1 x $$

La regresión polinómial es una técnica estadística que extiende la regresión lineal al incluir potencias de la variable independiente x, lo que permite capturar relaciones no lineales con la variable dependiente y. Aunque el modelo describe curvas, sigue siendo lineal en sus parámetros y por ello se considera un caso especial de regresión lineal múltiple. Su objetivo es encontrar la mejor curva de ajuste, generalmente mediante mínimos cuadrados, para explicar fenómenos donde una recta no basta. Además, se aplican pruebas para elegir el grado adecuado del polinomio y se utilizan herramientas como intervalos de confianza, bondad de ajuste y análisis de residuos para validar el modelo. (Conzmr, 2017)

$$ y = \beta_0 + \beta_1 x + \beta_2 x^2 + \beta_3 x^3 + \cdots + \beta_n x^n $$

La interacción ocurre cuando el efecto de una variable explicativa sobre el resultado depende del nivel de otra variable. En otras palabras, dos o más factores no se suman de forma independiente, sino que se alteran mutuamente. Por ejemplo, el efecto de X sobre Y puede variar si otra variable Z toma distintos valores. En regresión se modela con un término multiplicativo (por ejemplo X*Z) y, si su coeficiente es significativo, eso indica que hay una interacción real entre esos factores. (Camacho, s.f)

La significancia de factores se refiere a la comprobación estadística de si cada variable explicativa tiene un efecto real sobre la variable dependiente. En regresión, esto se hace mediante pruebas t, obteniendo p_values que indican la probabilidad de que ese efecto pueda ser cero por azar. Si el p-value es menor que un nivel de confianza, decimos que el factor es significativo. Esta evaluación ayuda a decidir qué variables quedan en el modelo final. (ChReInvent, s. f.)

La regularización es una técnica para mejorar modelos de regresión evitando que los coeficientes se vuelvan demasiado grandes (lo que puede provocar sobreajuste). 
- Ridge (o regularización L2) agrega un término que penaliza la suma de los cuadrados de los coeficientes. Así los coeficientes grandes se reducen, pero ninguno llega a cero. (IBM, 2023)
- Lasso (o regularización L1) penaliza la suma de los valores absolutos de los coeficientes. Esto puede forzar algunos coeficientes exactamente a cero, lo que equivale a una selección automática de variables. (Amat Rodrigo, 2016)
-  ElasticNet combina ambas penalizaciones (L1 + L2). De esta forma hereda ventajas de Ridge (coeficientes estables frente a correlación) y de Lasso (puede poner coeficientes a cero). Se usa cuando hay muchas variables correlacionadas. (Interactive Chaos, s. f.)

El tema elegido para proyecto se centra en evaluar la relación entre diferentes variables y las ventas de una empresa que se dedica a vender bicicletas mediante modelos de regresión lineal y sus variantes regularizadas (Ridge, Lasso y Elastic Net). El objetivo es identificar qué factores influyen de manera significativa en las ventas y cómo cambia el desempeño del modelo al aplicar distintas técnicas de regularización.

---
## Análisis de dataset

**¿De dónde viene?**

El dataset proviene de la página de internet Kaggle, la cual es una página de internet donde se pueden encontrar muchos datasets para el análisis de datos, este dataset viene de un repositorio que se llama “Company Financials Dataset” creado por Atharva Arya

**¿Qué contiene?**

El dataset contiene datos de ventas y costos simulados de una empresa durante el año 2024, contiene información financiera de ventas de productos incluyendo lo siguiente:

* Segmento de mercado

* País de venta

* Nombre y tipo del producto

* Precio de manufactura y precio de venta

* Unidades vendidas

* Ventas brutas y netas

* Costos de producción

* Utilidades

* Fechas

**¿Qué información dan las muestras?**

Cada fila del dataset es una muestra de transacción de ventas donde se puede ver el comportamiento del producto vendido en un país determinado, segmento y periodo, todo esto nos muestra el nivel de ventas alcanzado, la rentabilidad de cada producto, la relación entre lo vendido, precio, costos y utilidades

**¿Qué se quiere analizar?**

El objetivo es predecir y analizar las ventas de la empresa, entonces se requiere identificar qué factores son los que influyen más en las ventas, evaluar cómo cambian las utilidades según las variaciones en las variables y comparar los diferentes modelos de regresión para ver cuál es el que ofrece la mejor predicción

**¿Qué variables se tienen que transformar para poder usarse en un modelo de regresión?**

Primero se tienen que limpiar las variables numéricas que tienen caracteres especiales para poder convertirlos en valores numéricos y después transformar las variables categóricas, como el segmento de mercado, tipo de producto, país y el nombre del mes, transformarlas en variables dummies para poder hacer la regresión. También se van a tener que escalar o estandarizar los datos para mejorar la comparabilidad de los datos en los modelos.

In [13]:
import pandas as pd

data = pd.read_excel('DataSet_Proyecto.xlsx')
# Quitar espacios extra en los nombres de columnas
data.columns = data.columns.str.strip()
data

Segment                   Country      Product Discount Band  \
0          Government                    Canada   Carretera          None    
1          Government                   Germany   Carretera          None    
2           Midmarket                    France   Carretera          None    
3           Midmarket                   Germany   Carretera          None    
4           Midmarket                    Mexico   Carretera          None    
..                ...                       ...          ...           ...   
695    Small Business                    France    Amarilla          High    
696    Small Business                    Mexico    Amarilla          High    
697        Government                    Mexico     Montana          High    
698        Government                    Canada       Paseo          High    
699  Channel Partners  United States of America         VTT          High    

     Units Sold  Manufacturing Price  Sale Price     Gross Sales  \
0        1618.5                    3          20           32370   
1        1321.0                    3          20           26420   
2        2178.0                    3          15           32670   
3         888.0                    3          15           13320   
4        2470.0                    3          15           37050   
..          ...                  ...         ...             ...   
695      2475.0                  260         300   $7,42,500.00    
696       546.0                  260         300   $1,63,800.00    
697      1368.0                    5           7            9576   
698       723.0                   10           7            5061   
699      1806.0                  250          12           21672   

          Discounts           Sales            COGS   Profit       Date  \
0                 0           32370           16185    16185 2014-01-01   
1                 0           26420           13210    13210 2014-01-01   
2                 0           32670           21780    10890 2014-06-01   
3                 0           13320            8880     4440 2014-06-01   
4                 0           37050           24700    12350 2014-06-01   
..              ...             ...             ...      ...        ...   
695   $1,11,375.00    $6,31,125.00    $6,18,750.00     12375 2014-03-01   
696           24570   $1,39,230.00    $1,36,500.00      2730 2014-10-01   
697          1436.4          8139.6            6840   1299.6 2014-02-01   
698          759.15         4301.85            3615   686.85 2014-04-01   
699          3250.8         18421.2            5418  13003.2 2014-05-01   

     Month Number  Month Name  Year  
0               1    January   2014  
1               1    January   2014  
2               6       June   2014  
3               6       June   2014  
4               6       June   2014  
..            ...         ...   ...  
695             3      March   2014  
696            10    October   2014  
697             2   February   2014  
698             4      April   2014  
699             5        May   2014  

[700 rows x 16 columns]

In [14]:
# Quitar columnas irrelevantes para predecir nuestras ventas
data = data.drop(columns=['Discount Band','Date','Month Name'])
data

Segment                   Country      Product  Units Sold  \
0          Government                    Canada   Carretera       1618.5   
1          Government                   Germany   Carretera       1321.0   
2           Midmarket                    France   Carretera       2178.0   
3           Midmarket                   Germany   Carretera        888.0   
4           Midmarket                    Mexico   Carretera       2470.0   
..                ...                       ...          ...         ...   
695    Small Business                    France    Amarilla       2475.0   
696    Small Business                    Mexico    Amarilla        546.0   
697        Government                    Mexico     Montana       1368.0   
698        Government                    Canada       Paseo        723.0   
699  Channel Partners  United States of America         VTT       1806.0   

     Manufacturing Price  Sale Price     Gross Sales       Discounts  \
0                      3          20           32370               0   
1                      3          20           26420               0   
2                      3          15           32670               0   
3                      3          15           13320               0   
4                      3          15           37050               0   
..                   ...         ...             ...             ...   
695                  260         300   $7,42,500.00    $1,11,375.00    
696                  260         300   $1,63,800.00            24570   
697                    5           7            9576          1436.4   
698                   10           7            5061          759.15   
699                  250          12           21672          3250.8   

              Sales            COGS   Profit  Month Number  Year  
0             32370           16185    16185             1  2014  
1             26420           13210    13210             1  2014  
2             32670           21780    10890             6  2014  
3             13320            8880     4440             6  2014  
4             37050           24700    12350             6  2014  
..              ...             ...      ...           ...   ...  
695   $6,31,125.00    $6,18,750.00     12375             3  2014  
696   $1,39,230.00    $1,36,500.00      2730            10  2014  
697          8139.6            6840   1299.6             2  2014  
698         4301.85            3615   686.85             4  2014  
699         18421.2            5418  13003.2             5  2014  

[700 rows x 13 columns]

In [15]:
# Quitamos las comas y signos de las columnas
cols_a_convertir = ['Units Sold', 'Manufacturing Price', 'Sale Price', 'Gross Sales','Sales','COGS','Profit','Discounts']

# Reemplazamos $ y comas en todas las columnas seleccionadas
data[cols_a_convertir] = data[cols_a_convertir].replace('[\$,]', '', regex=True).astype(float)
data

<>:5: SyntaxWarning: invalid escape sequence '\$'
<>:5: SyntaxWarning: invalid escape sequence '\$'
/var/folders/20/xvvxf3fd0bsc8mhjgvpylzzc0000gn/T/ipykernel_25066/3704823260.py:5: SyntaxWarning: invalid escape sequence '\$'
  data[cols_a_convertir] = data[cols_a_convertir].replace('[\$,]', '', regex=True).astype(float)


Segment                   Country      Product  Units Sold  \
0          Government                    Canada   Carretera       1618.5   
1          Government                   Germany   Carretera       1321.0   
2           Midmarket                    France   Carretera       2178.0   
3           Midmarket                   Germany   Carretera        888.0   
4           Midmarket                    Mexico   Carretera       2470.0   
..                ...                       ...          ...         ...   
695    Small Business                    France    Amarilla       2475.0   
696    Small Business                    Mexico    Amarilla        546.0   
697        Government                    Mexico     Montana       1368.0   
698        Government                    Canada       Paseo        723.0   
699  Channel Partners  United States of America         VTT       1806.0   

     Manufacturing Price  Sale Price  Gross Sales  Discounts      Sales  \
0                    3.0        20.0      32370.0       0.00   32370.00   
1                    3.0        20.0      26420.0       0.00   26420.00   
2                    3.0        15.0      32670.0       0.00   32670.00   
3                    3.0        15.0      13320.0       0.00   13320.00   
4                    3.0        15.0      37050.0       0.00   37050.00   
..                   ...         ...          ...        ...        ...   
695                260.0       300.0     742500.0  111375.00  631125.00   
696                260.0       300.0     163800.0   24570.00  139230.00   
697                  5.0         7.0       9576.0    1436.40    8139.60   
698                 10.0         7.0       5061.0     759.15    4301.85   
699                250.0        12.0      21672.0    3250.80   18421.20   

         COGS    Profit  Month Number  Year  
0     16185.0  16185.00             1  2014  
1     13210.0  13210.00             1  2014  
2     21780.0  10890.00             6  2014  
3      8880.0   4440.00             6  2014  
4     24700.0  12350.00             6  2014  
..        ...       ...           ...   ...  
695  618750.0  12375.00             3  2014  
696  136500.0   2730.00            10  2014  
697    6840.0   1299.60             2  2014  
698    3615.0    686.85             4  2014  
699    5418.0  13003.20             5  2014  

[700 rows x 13 columns]

In [16]:
data.describe()

Units Sold  Manufacturing Price  Sale Price   Gross Sales  \
count   700.000000           700.000000  700.000000  7.000000e+02   
mean   1608.294286            96.477143  118.428571  1.827594e+05   
std     867.427859           108.602612  136.775515  2.542623e+05   
min     200.000000             3.000000    7.000000  1.799000e+03   
25%     905.000000             5.000000   12.000000  1.739175e+04   
50%    1542.500000            10.000000   20.000000  3.798000e+04   
75%    2229.125000           250.000000  300.000000  2.790250e+05   
max    4492.500000           260.000000  350.000000  1.207500e+06   

           Discounts         Sales           COGS         Profit  \
count     700.000000  7.000000e+02     700.000000     700.000000   
mean    13150.354671  1.696091e+05  145475.211429   24133.860414   
std     22962.928760  2.367263e+05  203865.506118   42760.626547   
min         0.000000  1.655080e+03     918.000000  -40617.500000   
25%       800.320000  1.592800e+04    7490.000000    2805.960000   
50%      2585.250000  3.554020e+04   22506.250000    9242.200000   
75%     15956.347500  2.610775e+05  245607.500000   22662.000000   
max    149677.500000  1.159200e+06  950625.000000  262200.000000   

       Month Number         Year  
count    700.000000   700.000000  
mean       7.900000  2013.750000  
std        3.377321     0.433322  
min        1.000000  2013.000000  
25%        5.750000  2013.750000  
50%        9.000000  2014.000000  
75%       10.250000  2014.000000  
max       12.000000  2014.000000

In [17]:
# Analizamos si nos conviene convertir la variable de países a una de región
data['Country'].value_counts()

Country
Canada                      140
Germany                     140
France                      140
Mexico                      140
United States of America    140
Name: count, dtype: int64

In [18]:
# Podemos seguir trabajando con países, ya que no hay muchos, no es necesario agruparlos por regiones
# Ahora continuamos con la preparación de nuestros datos, convertimos las variables categóricas Segment, Country, Product a numéricas (dummies)
data = pd.get_dummies(data, columns=['Segment', 'Country', 'Product'], drop_first=True).astype(float)
data.head()

Units Sold  Manufacturing Price  Sale Price  Gross Sales  Discounts  \
0      1618.5                  3.0        20.0      32370.0        0.0   
1      1321.0                  3.0        20.0      26420.0        0.0   
2      2178.0                  3.0        15.0      32670.0        0.0   
3       888.0                  3.0        15.0      13320.0        0.0   
4      2470.0                  3.0        15.0      37050.0        0.0   

     Sales     COGS   Profit  Month Number    Year  ...  \
0  32370.0  16185.0  16185.0           1.0  2014.0  ...   
1  26420.0  13210.0  13210.0           1.0  2014.0  ...   
2  32670.0  21780.0  10890.0           6.0  2014.0  ...   
3  13320.0   8880.0   4440.0           6.0  2014.0  ...   
4  37050.0  24700.0  12350.0           6.0  2014.0  ...   

   Segment_Small Business  Country_France  Country_Germany  Country_Mexico  \
0                     0.0             0.0              0.0             0.0   
1                     0.0             0.0              1.0             0.0   
2                     0.0             1.0              0.0             0.0   
3                     0.0             0.0              1.0             0.0   
4                     0.0             0.0              0.0             1.0   

   Country_United States of America  Product_ Carretera   Product_ Montana   \
0                               0.0                  1.0                0.0   
1                               0.0                  1.0                0.0   
2                               0.0                  1.0                0.0   
3                               0.0                  1.0                0.0   
4                               0.0                  1.0                0.0   

   Product_ Paseo   Product_ VTT   Product_ Velo   
0              0.0            0.0             0.0  
1              0.0            0.0             0.0  
2              0.0            0.0             0.0  
3              0.0            0.0             0.0  
4              0.0            0.0             0.0  

[5 rows x 23 columns]

In [19]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 700 entries, 0 to 699
Data columns (total 23 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Units Sold                        700 non-null    float64
 1   Manufacturing Price               700 non-null    float64
 2   Sale Price                        700 non-null    float64
 3   Gross Sales                       700 non-null    float64
 4   Discounts                         700 non-null    float64
 5   Sales                             700 non-null    float64
 6   COGS                              700 non-null    float64
 7   Profit                            700 non-null    float64
 8   Month Number                      700 non-null    float64
 9   Year                              700 non-null    float64
 10  Segment_Enterprise                700 non-null    float64
 11  Segment_Government                700 non-null    float64
 12  Segment_

---
## Modelo propuesto
Los 3 modelos distintos para comparar entre ellos serán:

- 1. Lineal Múltiple (relación directa de costos y ventas)
- 2. Polinomial (grado 2)
- 3. Lineal Múltiple usando Precio Efectivo (variale obtenida)

Cada modelo tendrá 4 versiones: sin penalización, Ridge, Lasso, y ElasticNet.

---
### 1. Lineal Múltiple (relación directa de costos y ventas)

### Sin Penalización

In [20]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
import statsmodels.api as sm
from scipy import stats
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Ridge, Lasso, ElasticNet

In [21]:
y = data["Sales"]
x = data[["COGS", "Discounts", "Units Sold", "Manufacturing Price"]] 

In [22]:
n = len(data)

In [23]:
x_train, x_test, y_train, y_test = train_test_split(x, y, train_size = 0.4, random_state = 137)

#### Escalamiento

In [24]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler().fit(x_train)
X_train_scaled = scaler.transform(x_train)
lr = LinearRegression()
lr.fit(X_train_scaled, y_train)

LinearRegression()

#### Train

In [25]:
y_hat = lr.predict(X_train_scaled)
r2_score(y_train, y_hat)

0.9884871763859922

In [26]:
n = len(X_train_scaled)
unos = np.ones([n,1])
X = np.column_stack([unos,X_train_scaled])
ols = sm.OLS(y_train, X)
results = ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.988
Model:                            OLS   Adj. R-squared:                  0.988
Method:                 Least Squares   F-statistic:                     5903.
Date:                Sun, 28 Sep 2025   Prob (F-statistic):          3.54e-265
Time:                        17:09:24   Log-Likelihood:                -3243.8
No. Observations:                 280   AIC:                             6498.
Df Residuals:                     275   BIC:                             6516.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       1.777e+05   1568.235    113.343      0.000    1.75e+05    1.81e+05
x1          2.606e+05   2746.076     94.889      0.000    2.55e+05    2.66e+05
x2          -2.51e+04   2690.677     -9.330      0.000   -3.04e+04   -1.98e+04
x3           980.9174   1676.315      0.585      0.559   -2319.123    4280.958
x4           178.3495   1574.536      0.113      0.910   -2921.327    3278.026
==============================================================================
Omnibus:                       20.131   Durbin-Watson:                   2.061
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               27.713
Skew:                          -0.514   Prob(JB):                     9.60e-07
Kurtosis:                       4.149   Cond. No.                         3.30
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Los p-values de COGS y Discounts son de 0 por lo que sí influyen en las ventas, COGS alza ventas y descuentos las bajan. Los p-values de Units Sold y Manufacturing Price son grandes por lo que no hay evidencia de que aporten algo extra cuando COGS ya está en el modelo. 

#### Interpretación signos betas

- const: Las ventas cuando las demás variables están en su nivel promedio son de aproximadamente 177,700
- COGS: Si el Cost of Groods Sold sube las ventas incrementan en más de 260,600
- Discounts: Si los descuentos suben las ventas bajan en aproximadamente 25,100
- Units Sold: Cuando las unidades vendidas incrementan las ventas también lo hacen en 980.9174
- Manufacturing Price: Cuando el precio de manufactura incrementa, las ventas lo hacen en 178.3495

#### Test

In [27]:
X_test_scaled = scaler.transform(x_test)
lr = LinearRegression()
lr.fit(X_test_scaled, y_test)

LinearRegression()

In [28]:
y_hat_test = lr.predict(X_test_scaled)
r2_score(y_test, y_hat_test)

0.9871901733701929

In [29]:
n = len(X_test_scaled)
unos = np.ones([n,1])
X = np.column_stack([unos,X_test_scaled])
ols = sm.OLS(y_test, X)
results = ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.987
Model:                            OLS   Adj. R-squared:                  0.987
Method:                 Least Squares   F-statistic:                     7996.
Date:                Sun, 28 Sep 2025   Prob (F-statistic):               0.00
Time:                        17:09:24   Log-Likelihood:                -4870.6
No. Observations:                 420   AIC:                             9751.
Df Residuals:                     415   BIC:                             9771.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       1.752e+05   1306.638    134.068      0.000    1.73e+05    1.78e+05
x1          2.498e+05   2098.769    119.011      0.000    2.46e+05    2.54e+05
x2         -1.838e+04   2016.892     -9.115      0.000   -2.23e+04   -1.44e+04
x3         -1356.0176   1249.710     -1.085      0.279   -3812.568    1100.533
x4          1328.9468   1276.174      1.041      0.298   -1179.625    3837.518
==============================================================================
Omnibus:                       34.864   Durbin-Watson:                   1.910
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              150.449
Skew:                           0.116   Prob(JB):                     2.14e-33
Kurtosis:                       5.923   Cond. No.                         2.96
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Al igual que el modelo con Train, en el conjunto de Test, COGS y Discounts son claramente significativos (p-values < 0.001) pues COGS empuja ventas hacia arriba y los descuentos las reducen. En cambio, Units Sold y Manufacturing Price no son significativos al tener COGS en el modelo.

#### Interpretación signos Betas

- const: Ventas promedio cuando las demás variables están en su nivel promedio son de aproximadamente 175,200.
- COGS: Si el COGS sube, las ventas aumentan aproximadamente 249,800.
- Discounts: Si los descuentos suben, las ventas bajan aproximadamente 18,400.
- Units Sold: Tiene un efecto pequeño, no vemos cambio claro en ventas (-1356.0176).
- Manufacturing Price: Tiene un efecto pequeño, tampoco hay un cambio claro (1328.9468)

### Ridge

In [30]:
l2 = Ridge(alpha = 0.1)  
l2.fit(x, y)

Ridge(alpha=0.1)

In [31]:
y_pred = l2.predict(x)

In [32]:
r2_score(y, y_pred)

0.9874531273335674

In [33]:
l2.intercept_

2691.460185948963

In [34]:
l2.coef_

array([ 1.2315188 , -0.88895857, -0.80363359,  7.71947723])

In [35]:
l2 = Ridge(alpha = 1)  
l2.fit(x, y)
y_pred = l2.predict(x)
r2_score(y, y_pred)

0.9874531273335674

In [36]:
l2.intercept_

2691.4602700232354

In [37]:
l2.coef_

array([ 1.2315188 , -0.88895857, -0.8036336 ,  7.71947639])

In [38]:
l2 = Ridge(alpha = 10)  
l2.fit(x, y)
y_pred = l2.predict(x)
r2_score(y, y_pred)

0.9874531273335674

In [39]:
l2.intercept_

2691.4611107650853

In [40]:
l2.coef_

array([ 1.2315188 , -0.88895857, -0.80363364,  7.71946792])

#### Interpretación

Al comparar los modelos de Ridge con distintos valores de alpha, se observa que el desempeño medido con el R2 prácticamente no cambia, lo cual significa que el ajuste del modelo sigue siendo igual de bueno. Lo que sí cambia, aunque de manera muy ligera, son los coeficientes, al aumentar el valor de alpha, los coeficientes se hacen un poco más pequeños y estables, pero sin modificar de forma importante el resultado final. Esto refleja que Ridge introduce un pequeño ajuste para evitar que los coeficientes crezcan demasiado, pero en este caso el efecto es mínimo porque el modelo ya estaba bastante bien ajustado

### Ridge Train

In [41]:
l2 = Ridge(alpha = 0.1)  
l2.fit(X_train_scaled, y_train)

Ridge(alpha=0.1)

In [42]:
y_pred = l2.predict(X_train_scaled)
r2_score(y_train, y_pred)

0.9884866549688349

In [43]:
l2.intercept_

177747.69771428572

In [44]:
l2.coef_

array([ 2.60267435e+05, -2.48556138e+04,  1.01520058e+03,  1.88464176e+02])

In [45]:
l2 = Ridge(alpha = 1)  
l2.fit(X_train_scaled, y_train)

Ridge(alpha=1)

In [46]:
l2.intercept_

177747.69771428572

In [47]:
l2.coef_

array([257561.87060672, -22672.87083967,   1319.04616626,    277.49357921])

In [48]:
l2 = Ridge(alpha = 10)  
l2.fit(X_train_scaled, y_train)
y_pred = l2.predict(X_train_scaled)
r2_score(y_train, y_pred)

0.984726110884884

In [49]:
l2.intercept_

177747.69771428572

In [50]:
l2.coef_

array([234549.35852893,  -4652.18597342,   3959.54073029,   1001.99543288])

#### Interpretación

Con los R2 se puede ver que el valor se mantiene prácticamente igual en los tres casos 0.9884 con alpha = 0.1, 0.9884 con alpha = 1 y 0.9847 con alpha = 10. Esto significa que el cambio en alpha no afecta mucho la capacidad del modelo para explicar las ventas, ya que la explicación de la variabilidad de los datos sigue siendo muy alta. Sin embargo, cuando el alpha sube a 10, el R2 baja un poco, lo que indica que el modelo empieza a perder algo de ajuste en favor de tener coeficientes más pequeños y un modelo más estable.

### Ridge test

In [51]:
l2 = Ridge(alpha = 0.1)
l2.fit(X_test_scaled, y_test)

Ridge(alpha=0.1)

In [52]:
y_pred_test = l2.predict(X_test_scaled)
r2_score(y_test, y_pred_test)

0.9871899815353641

In [53]:
l2.intercept_

175178.03637198082

In [54]:
l2.coef_

array([249610.71766933, -18260.96035356,  -1333.54697641,   1334.56429711])

In [55]:
l2 = Ridge(alpha = 1)
l2.fit(X_test_scaled, y_test)
y_pred_test = l2.predict(X_test_scaled)
r2_score(y_test, y_pred_test)

0.9871713260825816

In [56]:
l2.intercept_

175180.61143422438

In [57]:
l2.coef_

array([248137.45892837, -17165.45499094,  -1133.15470005,   1384.58880036])

In [58]:
l2 = Ridge(alpha = 10)
l2.fit(X_test_scaled, y_test)
y_pred_test = l2.predict(X_test_scaled)
r2_score(y_test, y_pred_test)

0.9855956340350267

In [59]:
l2.intercept_

175185.7115060192

In [60]:
l2.coef_

array([234694.66756886,  -7374.03460377,    704.36241025,   1836.96488924])

#### Interpretación

En el modelo Ridge con los datos de prueba, se observa que al usar un alpha pequeño (0.1) los coeficientes son más grandes y el R2 es muy alto (0.9871). Cuando se aumenta el alpha a 1 y luego a 10, los coeficientes se van reduciendo en magnitud, mientras que el R2 apenas baja un poco (de 0.9871 a 0.9859). Esto significa que, aunque el modelo sigue explicando muy bien las ventas, con un mayor alpha los coeficientes quedan más controlados y el riesgo de sobreajuste disminuye

### Lasso

In [61]:
l1 = Lasso(alpha = 0.1)   
l1.fit(x, y)
y_pred = l1.predict(x)
r2_score(y, y_pred)

0.9874531273335674

In [62]:
l1.intercept_

2691.4608302671113

In [63]:
l1.coef_

array([ 1.2315188 , -0.88895857, -0.8036335 ,  7.71946885])

In [64]:
l1 = Lasso(alpha = 1)   
l1.fit(x, y)
y_pred = l1.predict(x)
r2_score(y, y_pred)

0.9874531273335658

In [65]:
l1.intercept_

2691.466713203292

In [66]:
l1.coef_

array([ 1.23151881, -0.88895858, -0.80363264,  7.71939255])

In [67]:
l1 = Lasso(alpha = 10)   
l1.fit(x, y)
y_pred = l1.predict(x)
r2_score(y, y_pred)

0.9874531273334142

In [68]:
l1.intercept_

2691.525542546733

In [69]:
l1.coef_

array([ 1.23151882, -0.88895868, -0.80362403,  7.71862952])

#### Interpretación

En el modelo Lasso, al cambiar el valor de alpha (0.1, 1 y 10) prácticamente no hay un impacto grande en el ajuste del modelo, el R² se mantiene casi igual, 0.987 en todos los casos, y los coeficientes solo tienen variaciones muy pequeñas. Esto pasa porque tu modelo ya está muy bien ajustado y la penalización de Lasso, que normalmente reduce coeficientes hacia cero, apenas tiene efecto

### Lasso Train

In [70]:
l1 = Lasso(alpha = 0.1)
l1.fit(X_train_scaled, y_train)

Lasso(alpha=0.1)

In [71]:
y_pred = l1.predict(X_train_scaled)
r2_score(y_train, y_pred)

0.9884871763837466

In [72]:
l1.intercept_

177747.69771428572

In [73]:
l1.coef_

array([ 2.60572925e+05, -2.51026947e+04,  9.80834721e+02,  1.78259761e+02])

In [74]:
l1 = Lasso(alpha = 1)
l1.fit(X_train_scaled, y_train)
y_pred = l1.predict(X_train_scaled)
r2_score(y_train, y_pred)

0.988487176156903

In [75]:
l1.intercept_

177747.69771428572

In [76]:
l1.coef_

array([ 2.60568057e+05, -2.50976495e+04,  9.80100263e+02,  1.77454434e+02])

In [77]:
l1 = Lasso(alpha = 10)
l1.fit(X_train_scaled, y_train)
y_pred = l1.predict(X_train_scaled)
r2_score(y_train, y_pred)

0.9884871539743914

In [78]:
l1.intercept_

177747.69771428572

In [79]:
l1.coef_

array([ 2.60520297e+05, -2.50478991e+04,  9.72635239e+02,  1.69370066e+02])

#### Interpretación

Con un alpha pequeño 0.1, los coeficientes se mantienen más grandes porque el modelo casi no los reduce. A medida que subes a alpha = 1 o 10, el modelo empieza a encoger más los coeficientes, haciéndolos más pequeños en magnitud. Esto significa que el modelo se vuelve más simple porque reduce la importancia de algunas variables. Sin embargo, el R2 se mantiene casi igual (0.988), lo que indica que el ajuste sigue siendo muy bueno y el cambio de alpha no afecta mucho la capacidad de predicción en este caso, pero sí influye en la fuerza con que se penalizan los coeficientes.

### Lasso Test

In [80]:
l1 = Lasso(alpha = 0.1)
l1.fit(X_test_scaled, y_test)

Lasso(alpha=0.1)

In [81]:
y_pred_test = l1.predict(X_test_scaled)
r2_score(y_test, y_pred_test)

0.9871901733679374

In [82]:
l1.intercept_

175177.718450723

In [83]:
l1.coef_

array([249775.50741897, -18383.67925253,  -1355.88258514,   1328.86497025])

In [84]:
l1 = Lasso(alpha = 1)
l1.fit(X_test_scaled, y_test)
y_pred_test = l1.predict(X_test_scaled)
r2_score(y_test, y_pred_test)

0.9871901731369351

In [85]:
l1.intercept_

175177.6668790574

In [86]:
l1.coef_

array([249770.85390461, -18379.54340528,  -1354.65113313,   1328.13236101])

In [87]:
l1 = Lasso(alpha = 10)
l1.fit(X_test_scaled, y_test)
y_pred_test = l1.predict(X_test_scaled)
r2_score(y_test, y_pred_test)

0.9871901504223625

In [88]:
l1.intercept_

175177.15360333084

In [89]:
l1.coef_

array([249724.95991578, -18338.625528  ,  -1342.43072656,   1320.78165814])

#### Interpretación

En este caso con Lasso Test, al variar el valor de alpha (0.1, 1 y 10) el modelo prácticamente mantiene el mismo nivel de ajuste (el R2 se queda en 0.987). Lo que cambia un poco son los coeficientes, a medida que sube el alpha, los valores de los betas se van reduciendo ligeramente en magnitud, acercándose más a cero.

### ElasticNet

In [90]:
le = ElasticNet(alpha=0.1)
le.fit(x, y)

ElasticNet(alpha=0.1)

In [91]:
y_pred = le.predict(x)

In [92]:
r2_score(y, y_pred)

0.987453127333567

In [93]:
le.intercept_

2691.463772977004

In [94]:
le.coef_

array([ 1.23151881, -0.88895858, -0.8036337 ,  7.71944017])

In [95]:
le = ElasticNet(alpha=1)
le.fit(x, y)
y_pred = le.predict(x)
r2_score(y, y_pred)

0.9874531273335384

In [96]:
le.intercept_

2691.496138847957

In [97]:
le.coef_

array([ 1.23151882, -0.88895864, -0.80363462,  7.71910576])

In [98]:
le = ElasticNet(alpha=10)
le.fit(x, y)
y_pred = le.predict(x)
r2_score(y, y_pred)

0.9874531273306767

In [99]:
le.intercept_

2691.8196535649186

In [100]:
le.coef_

array([ 1.23151897, -0.88895925, -0.80364383,  7.71576305])

#### Interpretación

En el modelo de Elastic Net con diferentes valores de alpha (0.1, 1 y 10), se observa que el R2 se mantiene prácticamente igual (0.987), lo que significa que el ajuste del modelo no cambia mucho en términos de capacidad predictiva. Sin embargo, al aumentar el valor de alpha, los coeficientes se van reduciendo ligeramente en magnitud, lo cual indica que Elastic Net está aplicando más regularización, combinando las ideas de Ridge (reducir coeficientes grandes) y Lasso (forzar coeficientes hacia cero)

### ElasticNet Train

In [101]:
le = ElasticNet(alpha = 0.1)
le.fit(X_train_scaled, y_train)
y_pred = le.predict(X_train_scaled)
r2_score(y_train, y_pred)

0.9819218628037095

In [102]:
le.intercept_

177747.69771428572

In [103]:
le.coef_

array([226187.76623471,   1610.29900415,   4944.35461443,   1247.91236182])

In [104]:
le = ElasticNet(alpha = 1)
le.fit(X_train_scaled, y_train)
y_pred = le.predict(X_train_scaled)
r2_score(y_train, y_pred)

0.8801636918017546

In [105]:
le.intercept_

177747.69771428572

In [106]:
le.coef_

array([128510.02651953,  51826.37309965,  16326.87967525,   2720.594907  ])

In [107]:
le = ElasticNet(alpha = 10)
le.fit(X_train_scaled, y_train)
y_pred = le.predict(X_train_scaled)
r2_score(y_train, y_pred)

0.42079696327792715

In [108]:
le.intercept_

177747.69771428572

In [109]:
le.coef_

array([35987.42300894, 25747.36578807, 10649.07508629,   714.14108253])

#### Interpretación

Cuando probamos Elastic Net en el conjunto de entrenamiento, vimos que el valor de alpha cambia mucho el resultado. Con alpha = 0.1, el modelo tuvo un R2 = 0.9819, lo que significa que explica casi el 98% de la variación de las ventas, y los coeficientes quedaron en valores moderados. Al aumentar a alpha = 1, el R2 bajó a 0.8801, y los coeficientes crecieron un poco más. Finalmente, con alpha = 10, el R2 cayó fuertemente a 0.4208, lo que significa que el modelo ya casi no explica bien los datos, y además los coeficientes se inflaron mucho.

### ElasticNet Test

In [110]:
le = Lasso(alpha = 0.1)
le.fit(X_test_scaled, y_test)
y_pred_test = le.predict(X_test_scaled)
r2_score(y_test, y_pred_test)

0.9871901733679374

In [111]:
le.intercept_

175177.718450723

In [112]:
le.coef_

array([249775.50741897, -18383.67925253,  -1355.88258514,   1328.86497025])

In [113]:
le = Lasso(alpha = 1)
le.fit(X_test_scaled, y_test)
y_pred_test = le.predict(X_test_scaled)
r2_score(y_test, y_pred_test)

0.9871901731369351

In [114]:
le.intercept_

175177.6668790574

In [115]:
le.coef_

array([249770.85390461, -18379.54340528,  -1354.65113313,   1328.13236101])

In [116]:
le = Lasso(alpha = 10)
le.fit(X_test_scaled, y_test)
y_pred_test = le.predict(X_test_scaled)
r2_score(y_test, y_pred_test)

0.9871901504223625

In [117]:
le.intercept_

175177.15360333084

In [118]:
le.coef_

array([249724.95991578, -18338.625528  ,  -1342.43072656,   1320.78165814])

#### Interpretación

En el modelo ElasticNet Test, se probaron diferentes valores de alpha (0.1, 1 y 10) y los resultados muestran que el desempeño del modelo, medido con R2, se mantiene prácticamente igual en todos los casos. Esto quiere decir que el ajuste del modelo a los datos de prueba no cambia mucho al variar alpha. Lo que si cambiaron fueron los betas, por ejemplo, con un alpha pequeño los valores de los parámetros son un poco más grandes en magnitud, mientras que al aumentar alpha estos coeficientes se reducen ligeramente y se hacen más pequeños. Aumentar alpha hace que los coeficientes se reduzcan ligeramente en magnitud, pero sin afectar demasiado la capacidad predictiva del modelo.

---

### 2. Polinomial (grado 2)

---
### 3. Lineal Múltiple usando Precio Efectivo (variale obtenida)

In [134]:
#Preparar los datos para mi modelo
data_3 = data.copy()

# Columna precio efectivo
data_3['Precio_efectivo'] = data_3['Sale Price'] - (data_3['Discounts'] / data_3['Units Sold'])

# Columnas dummies
dummy_cols = []
for col in data_3.columns:
    if col.startswith(('Segment_', 'Country_', 'Product_')):
        dummy_cols.append(col)

y = data_3['Profit']
X = data_3[['Units Sold', 'Precio_efectivo'] + dummy_cols]

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size = 0.4, random_state = 137)

#### Regresiones utilizando los datos de entrenamiento

In [135]:
# Sin penalización

# Escalamiento
scaler = StandardScaler().fit(X_train)
X_scaled_train= scaler.transform(X_train)

n=len(X_scaled_train)
unos = np.ones([n,1])

X = np.hstack([unos, X_scaled_train])
ols = sm.OLS(y_train,X)
results = ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Profit   R-squared:                       0.763
Model:                            OLS   Adj. R-squared:                  0.750
Method:                 Least Squares   F-statistic:                     56.69
Date:                Sun, 28 Sep 2025   Prob (F-statistic):           1.89e-73
Time:                        18:14:52   Log-Likelihood:                -3199.5
No. Observations:                 280   AIC:                             6431.
Df Residuals:                     264   BIC:                             6489.
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       2.617e+04   1366.178     19.158      0.000    2.35e+04    2.89e+04
x1          1.296e+04   1446.333      8.958      0.000    1.01e+04    1.58e+04
x2          4.131e+04   1763.861     23.422      0.000    3.78e+04    4.48e+04
x3         -1.953e+04   1894.345    -10.312      0.000   -2.33e+04   -1.58e+04
x4         -4750.1886   2252.833     -2.109      0.036   -9185.995    -314.382
x5         -3625.3333   1796.309     -2.018      0.045   -7162.249     -88.418
x6         -2.077e+04   2206.906     -9.413      0.000   -2.51e+04   -1.64e+04
x7          -506.5224   1683.500     -0.301      0.764   -3821.318    2808.274
x8          1081.4579   1739.686      0.622      0.535   -2343.968    4506.884
x9         -1529.7429   1758.790     -0.870      0.385   -4992.784    1933.299
x10        -2835.1989   1744.915     -1.625      0.105   -6270.920     600.522
x11         -884.5498   1784.355     -0.496      0.621   -4397.929    2628.829
x12         -926.6090   1775.452     -0.522      0.602   -4422.458    2569.240
x13        -1182.5122   1946.416     -0.608      0.544   -5014.987    2649.962
x14         1195.0669   1775.030      0.673      0.501   -2299.951    4690.085
x15          212.5041   1892.859      0.112      0.911   -3514.517    3939.525
==============================================================================
Omnibus:                       85.892   Durbin-Watson:                   1.983
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              525.558
Skew:                           1.081   Prob(JB):                    7.52e-115
Kurtosis:                       9.354   Cond. No.                         3.42
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Interpretación Regresión sin penalización con datos de entrenamiento**

* El modelo tiene un R^2 de 0.763 el cual nos dice que hay una variabilidad del profit del 76.3%
* Units sold cuenta con un coeficiente positivo y con significancia estadistica, lo cual nos dice que a mayor volumen de ventas más profit
* El precio efectivo también cuenta con un coeficiente positivo y con significancia estadistica, lo cual nos dice que precios netos más altos, ajustados por descuentos, incrementan el Profit.
* Varios segmentos con significancia estadistica cuentan con coeficientes positivos lo cual nos dice que reducen el profic, segmentos como Enterprise, Government, Midmarket, Small Business reducen el Profit frente al segmento base Consumer.
* Las varriables de los paises no cuentan con signifciancia estadistica, lo cual nos dice que no afecta en el profit

In [141]:
# Con penalización Ridge
l2 = Ridge(alpha = 0.1)
l2.fit(X_scaled_train, y_train)
y_pred = l2.predict(X_scaled_train)
r2 = r2_score(y_train, y_pred)

print(f"R2: {r2}")
print(f"Constante: {l2.intercept_}")
print(f"Coeficiente: {l2.coef_}")

R2: 0.7631008110612786
Constante: 26173.29235714285
Coeficiente: [ 12949.65770228  41275.42041014 -19500.06439073  -4708.23589834
  -3607.91095061 -20723.30285169   -507.2792443    1079.00610192
  -1528.15779965  -2835.08640722   -880.45467747   -923.28484867
  -1181.07156767   1193.00015638    212.11039169]


**Interpretación Regresión con penalización Ridge (α = 0.1) con datos de entrenamiento**

* El modelo tiene un R² de 0.763, lo que significa que explica el 76.3% de la variabilidad del Profit, prácticamente igual que el modelo sin penalización.
* Units Sold mantiene un coeficiente positivo y fuerte, lo cual confirma que un mayor volumen de ventas incrementa significativamente el Profit.
* Precio efectivo también conserva un coeficiente positivo y elevado, mostrando que precios netos más altos (después de descuentos) aumentan el Profit.
 * Los segmentos (Enterprise, Government, Midmarket y Small Business) presentan coeficientes negativos, lo que indica que reducen el Profit frente al segmento base (Consumer). Ridge suavizó ligeramente estos valores, pero el efecto sigue siendo claro.
  * Las variables de países y productos fueron empujadas más cerca de cero y no muestran un efecto relevante en el Profit, lo que confirma que su influencia es poco significativa en este modelo.


In [142]:
# Con penalización Lasso
l1 = Lasso(alpha = 0.1)
l1.fit(X_scaled_train, y_train)
y_pred = l1.predict(X_scaled_train)
r2 = r2_score(y_train, y_pred)

print(f"R2: {r2}")
print(f"Constante: {l1.intercept_}")
print(f"Coeficiente: {l1.coef_}")

R2: 0.7631014326594544
Constante: 26173.29235714285
Coeficiente: [ 12955.75519359  41313.22553347 -19533.00529492  -4749.31312881
  -3624.81863301 -20772.2458942    -506.33696028   1081.43398179
  -1529.49533529  -2834.96647978   -884.26135517   -926.36844669
  -1182.28647016   1195.04028301    212.53204218]


**Interpretación Regresión con penalización Lasso (α = 0.1) con datos de entrenamiento**

* El modelo tiene un R² de 0.763, lo que significa que explica aproximadamente el 76.3% de la variabilidad del Profit, igual que en los modelos sin penalización y con Ridge.
* Units Sold mantiene un coeficiente positivo y fuerte, confirmando que un mayor volumen de ventas incrementa significativamente el Profit.
* El Precio efectivo conserva un coeficiente positivo y elevado, lo que indica que precios netos más altos (después de aplicar descuentos) aumentan el Profit.
* Los segmentos (Enterprise, Government, Midmarket y Small Business) presentan coeficientes negativos, reduciendo el Profit en comparación con el segmento base (Consumer).
* Las variables de países y productos muestran coeficientes más pequeños y cercanos a cero, lo que refleja que Lasso está reduciendo la relevancia de factores con poca contribución al modelo.
* A diferencia de Ridge, Lasso puede llevar a cero coeficientes irrelevantes si el α es más alto; en este caso (α = 0.1), los coeficientes se reducen pero aún no desaparecen por completo.


In [143]:
# Penalización elastic net
le = ElasticNet(alpha = 0.1)
le.fit(X_scaled_train, y_train)
y_pred = le.predict(X_scaled_train)
r2 = r2_score(y_train, y_pred)

print(f"R2: {r2}")
print(f"Constante: {le.intercept_}")
print(f"Coeficiente: {le.coef_}")

R2: 0.7554496103202067
Constante: 26173.29235714285
Coeficiente: [ 12175.42614909  36895.07797075 -15984.71130088   -475.20720913
  -1983.70966417 -15435.5491577    -542.04681855    857.61078214
  -1347.83836641  -2775.74546352   -517.76784366   -588.44323343
  -1049.43699113    946.35728316    107.04535165]


**Interpretación Regresión con penalización Elastic Net (α = 0.1) con datos de entrenamiento**

* El modelo tiene un R² de 0.755, lo que significa que explica aproximadamente el 75.5% de la variabilidad del Profit, ligeramente menor que en los modelos sin penalización, Ridge y Lasso.
* Units Sold mantiene un coeficiente positivo y fuerte, confirmando que el volumen de ventas incrementa de forma significativa el Profit.
* El Precio efectivo conserva un coeficiente positivo y elevado, mostrando que precios netos más altos, ajustados por descuentos, aumentan el Profit.
* Los segmentos (Enterprise, Government, Midmarket y Small Business) siguen mostrando coeficientes negativos, reduciendo el Profit en comparación con el segmento base (Consumer). Sin embargo, Elastic Net los suaviza aún más respecto a Lasso y Ridge.
* Las variables de países y productos son empujadas aún más cerca de cero, reforzando la idea de que su efecto sobre el Profit es poco relevante en este modelo.

### Regresiones utilizando datos de prueba


In [144]:
# Sin penalización

# Escalamiento
scaler = StandardScaler().fit(X_test)
X_scaled_test= scaler.transform(X_test)

n=len(X_scaled_test)
unos = np.ones([n,1])

X = np.hstack([unos, X_scaled_test])
ols = sm.OLS(y_test,X)
results = ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Profit   R-squared:                       0.699
Model:                            OLS   Adj. R-squared:                  0.688
Method:                 Least Squares   F-statistic:                     62.67
Date:                Sun, 28 Sep 2025   Prob (F-statistic):           2.03e-95
Time:                        18:57:48   Log-Likelihood:                -4800.8
No. Observations:                 420   AIC:                             9634.
Df Residuals:                     404   BIC:                             9698.
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       2.277e+04   1108.581     20.544      0.000    2.06e+04     2.5e+04
x1          1.224e+04   1129.408     10.836      0.000       1e+04    1.45e+04
x2          3.687e+04   1498.010     24.615      0.000    3.39e+04    3.98e+04
x3         -1.803e+04   1543.797    -11.678      0.000   -2.11e+04    -1.5e+04
x4         -2776.8668   1804.418     -1.539      0.125   -6324.088     770.354
x5         -2938.1331   1493.280     -1.968      0.050   -5873.703      -2.563
x6         -1.634e+04   1859.100     -8.787      0.000      -2e+04   -1.27e+04
x7          2163.0945   1445.130      1.497      0.135    -677.820    5004.009
x8          2798.3813   1445.749      1.936      0.054     -43.749    5640.511
x9          1076.5090   1422.056      0.757      0.449   -1719.044    3872.062
x10          491.5115   1410.660      0.348      0.728   -2281.639    3264.662
x11        -1484.8338   1506.898     -0.985      0.325   -4447.174    1477.506
x12         -892.9303   1506.781     -0.593      0.554   -3855.041    2069.181
x13          147.4676   1767.345      0.083      0.934   -3326.874    3621.809
x14        -1026.1179   1580.701     -0.649      0.517   -4133.545    2081.309
x15          221.6500   1506.786      0.147      0.883   -2740.471    3183.771
==============================================================================
Omnibus:                      160.775   Durbin-Watson:                   2.028
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1496.604
Skew:                           1.363   Prob(JB):                         0.00
Kurtosis:                      11.837   Cond. No.                         3.71
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Interpretación Regresión sin penalización con datos de prueba**

* El modelo tiene un R² de 0.699, lo que significa que explica aproximadamente el 69.9% de la variabilidad del Profit.
* Units Sold mantiene un coeficiente positivo y fuerte, confirmando que un mayor volumen de ventas incrementa el Profit.
* El Precio efectivo conserva un coeficiente positivo y elevado, mostrando que precios netos más altos, ajustados por descuentos, aumentan el Profit.
* Los segmentos Enterprise y Small Business presentan coeficientes negativos y significativos, lo que indica que reducen el Profit frente al segmento base (Consumer).
* Otros segmentos como Government y Midmarket se acercan al ser significativos, pero no son concluyentes.
* Las variables de países y productos no resultan significativas, lo que indica que no afectan de forma clara el Profit.

In [145]:
# Con penalización Ridge
l2 = Ridge(alpha = 0.1)
l2.fit(X_scaled_test, y_test)
y_pred = l2.predict(X_scaled_test)
r2 = r2_score(y_test, y_pred)

print(f"R2: {r2}")
print(f"Constante: {l2.intercept_}")
print(f"Coeficiente: {l2.coef_}")

R2: 0.6994266816139776
Constante: 22774.239119047626
Coeficiente: [ 12234.74049737  36849.60150939 -18009.11808716  -2755.11833667
  -2928.0567371  -16305.29081163   2161.90813173   2796.6679539
   1074.07876232    490.70426879  -1486.5103126    -896.08573745
    144.90544396  -1026.29702576    219.2515296 ]


**Interpretación Regresión con penalización Ridge (α = 0.1) con datos de prueba**

* El modelo tiene un R² de 0.699, prácticamente igual al modelo sin penalización, lo que indica que mantiene el mismo poder explicativo.
* Units Sold y Precio efectivo siguen siendo las variables más importantes con coeficientes positivos.
* Los segmentos continúan mostrando efectos negativos.
* Las variables de países y productos son casi cero, indicando que su efecto es poco relevante en la explicación del Profit.

In [146]:
# Con penalización Lasso
l1 = Lasso(alpha = 0.1)
l1.fit(X_scaled_test, y_test)
y_pred = l1.predict(X_scaled_test)
r2 = r2_score(y_test, y_pred)

print(f"R2: {r2}")
print(f"Constante: {l1.intercept_}")
print(f"Coeficiente: {l1.coef_}")

R2: 0.6994269552052952
Constante: 22774.239119047626
Coeficiente: [ 12238.62500893  36873.59245486 -18028.55497822  -2776.0982231
  -2937.62075037 -16334.75403773   2162.68476629   2797.9888783
   1076.06784812    491.11777829  -1484.75718658   -892.89701593
    147.38890942  -1026.0022913     221.53898875]


**Interpretación Regresión con penalización Lasso (α = 0.1) con datos de prueba**

* El modelo tiene un R² de 0.699, muy similar a OLS y Ridge, manteniendo el mismo poder explicativo.
Units Sold y Precio efectivo mantienen coeficientes positivos y fuertes, confirmando su significancia como factores clave del Profit.
* Los segmentos presentan coeficientes negativos, reduciendo el Profit en comparación con el segmento base (Consumer).
* Las variables de países y productos muestran coeficientes aún más pequeños, reflejando que Lasso penaliza con mayor fuerza a las variables de menor relevancia, aunque no las eliminó en este nivel de α.

In [147]:
# Penalización elastic net
le = ElasticNet(alpha = 0.1)
le.fit(X_scaled_test, y_test)
y_pred = le.predict(X_scaled_test)
r2 = r2_score(y_test, y_pred)

print(f"R2: {r2}")
print(f"Constante: {le.intercept_}")
print(f"Coeficiente: {le.coef_}")

R2: 0.691845281997308
Constante: 22774.239119047623
Coeficiente: [ 11450.99826003  32672.18890867 -14858.92547902    455.70906651
  -1551.78593048 -11517.33713779   1932.2364074    2458.71931801
    659.56778639    336.66909883  -1636.73025393  -1294.52394001
   -123.4831839    -932.50818717    -53.54663435]


**Interpretación Regresión con penalización Elastic Net (α = 0.1) con datos de prueba**

* El modelo tiene un R² de 0.691, lo que significa que explica aproximadamente el 69.1% de la variabilidad del Profit, ligeramente menor que en OLS, Ridge y Lasso.
* Units Sold mantiene un coeficiente positivo y fuerte, confirmando que el volumen de ventas incrementa de forma significativa el Profit.
* El Precio efectivo conserva un coeficiente positivo y elevado, mostrando que precios netos más altos, ajustados por descuentos, aumentan el Profit.
* Los segmentos (Enterprise, Government, Midmarket y Small Business) continúan mostrando coeficientes negativos, reduciendo el Profit en comparación con el segmento base (Consumer), aunque Elastic Net los suaviza más respecto a Ridge y Lasso.
* Las variables de países y productos se acercan aún más cerca a cero, reforzando que su impacto en el Profit es poco relevante dentro del modelo.

---
### Bibliografia APA

- IBM. (2021). What Is Linear Regression? Think. https://www.ibm.com/think/topics/linear-regression
- Conzmr. (2017). Regresión Polinomial. Recuperado de https://conzmr.wordpress.com/2017/04/04/regresion-polinomial/
- Camacho, C. (s. f.). Interacción en regresión [PDF]. Universidad de Sevilla. Recuperado de https://personal.us.es/vararey/interaccion1.pdf
- ChReInvent. (s. f.). Valor-p en regresión lineal. Recuperado de https://www.chreinvent.com/recursos/valor-p-en-regresi%C3%B3n-lineal
- IBM. (2023). ¿Qué es regression de ridge? Think. https://www.ibm.com/mx-es/think/topics/ridge-regression
- Amat Rodrigo, J. (2016). Selección de predictores, regularización ridge, lasso, elasticnet y reducción de dimensionalidad. Cienciadedatos.net. Recuperado de https://cienciadedatos.net/documentos/31_seleccion_de_predictores_subset_selection_ridge_lasso_dimension_reduction
- Interactive Chaos. (s. f.). Elastic Net. Recuperado de https://interactivechaos.com/es/manual/tutorial-de-machine-learning/elastic-net
- Company Financials Dataset. (2023, August 1). https://www.kaggle.com/datasets/atharvaarya25/financials/data